File: src/simplace/SQ_Energy_Balance/net/simplace/sim/components/SQ_Energy_Balance/CanopyTemperature.java
package net.simplace.sim.components.SQ_Energy_Balance;
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom2.Element;


public class CanopyTemperature extends FWSimComponent
{
    private FWSimVariable<Double> minTair;
    private FWSimVariable<Double> maxTair;
    private FWSimVariable<Double> cropHeatFlux;
    private FWSimVariable<Double> conductance;
    private FWSimVariable<Double> lambdaV;
    private FWSimVariable<Double> rhoDensityAir;
    private FWSimVariable<Double> specificHeatCapacityAir;
    private FWSimVariable<Double> minCanopyTemperature;
    private FWSimVariable<Double> maxCanopyTemperature;

    public CanopyTemperature(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public CanopyTemperature(){
        super();
    }
    @Override
    protected void process()
    {
        double t_minTair = minTair.getValue();
        double t_maxTair = maxTair.getValue();
        double t_cropHeatFlux = cropHeatFlux.getValue();
        double t_conductance = conductance.getValue();
        double t_lambdaV = lambdaV.getValue();
        double t_rhoDensityAir = rhoDensityAir.getValue();
        double t_specificHeatCapacityAir = specificHeatCapacityAir.getValue();
        double t_minCanopyTemperature;
        double t_maxCanopyTemperature;
        t_minCanopyTemperature = t_minTair + (t_cropHeatFlux / (t_rhoDensityAir * t_specificHeatCapacityAir * t_conductance / t_lambdaV * 1000.0d));
        t_maxCanopyTemperature = t_maxTair + (t_cropHeatFlux / (t_rhoDensityAir * t_specificHeatCapacityAir * t_conductance / t_lambdaV * 1000.0d));
        minCanopyTemperature.setValue(t_minCanopyTemperature, this);
        maxCanopyTemperature.setValue(t_maxCanopyTemperature, this);
    }

    @Override
    protected void init()
    {
    }
    public HashMap<String, FWSimVariable<?>> fillTestVariables(int aParamIndex, TEST_STATE aDefineOrCheck)
    {
        if(aParamIndex == 0 && aDefineOrCheck == TEST_STATE.DEFINE) //before process
        {
            FWSimVariable.setValue(1.225, iFieldMap.get("CanopyTemperature.rhoDensityAir"), this);
            FWSimVariable.setValue(0.7, iFieldMap.get("CanopyTemperature.minTair"), this);
            FWSimVariable.setValue(7.2, iFieldMap.get("CanopyTemperature.maxTair"), this);
        }
        else if(aParamIndex ==  0 && aDefineOrCheck == TEST_STATE.CHECK) //after process
        {
            FWSimVariable.setValue(2.184, iFieldMap.get("CanopyTemperature.minCanopyTemperature"), this);
            FWSimVariable.setValue(8.684, iFieldMap.get("CanopyTemperature.maxCanopyTemperature"), this);
        }
        else return null;
        return iFieldMap;
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new CanopyTemperature(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("minTair", "minimum air temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"", -30, 45, 0.7, this));
        addVariable(FWSimVariable.createSimVariable("maxTair", "maximum air Temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"", -30, 45, 7.2, this));
        addVariable(FWSimVariable.createSimVariable("cropHeatFlux", "Crop heat flux", DATA_TYPE.DOUBLE, CONTENT_TYPE.rate,"", 0, 10000, 447.912, this));
        addVariable(FWSimVariable.createSimVariable("conductance", "the boundary layer conductance", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"", 0, 10000, 598.685, this));
        addVariable(FWSimVariable.createSimVariable("lambdaV", "latent heat of vaporization of water", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 10, 2.454, this));
        addVariable(FWSimVariable.createSimVariable("rhoDensityAir", "Density of air", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", null, null, 1.225, this));
        addVariable(FWSimVariable.createSimVariable("specificHeatCapacityAir", "Specific heat capacity of dry air", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", null, null, 0.00101, this));
        addVariable(FWSimVariable.createSimVariable("minCanopyTemperature", "minimal Canopy Temperature  ", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"", -30, 45, null, this));
        addVariable(FWSimVariable.createSimVariable("maxCanopyTemperature", "maximal Canopy Temperature ", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"", -30, 45, null, this));

        return iFieldMap;
    }
}

File: src/cs/SQ_Energy_Balance/Canopytemperature.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class CanopyTemperature
{
    private double _lambdaV;
    public double lambdaV
        {
            get { return this._lambdaV; }
            set { this._lambdaV= value; } 
        }
    private double _rhoDensityAir;
    public double rhoDensityAir
        {
            get { return this._rhoDensityAir; }
            set { this._rhoDensityAir= value; } 
        }
    private double _specificHeatCapacityAir;
    public double specificHeatCapacityAir
        {
            get { return this._specificHeatCapacityAir; }
            set { this._specificHeatCapacityAir= value; } 
        }
    public CanopyTemperature() { }
    
    public void  CalculateModel(EnergyBalanceState s, EnergyBalanceState s1, EnergyBalanceRate r, EnergyBalanceAuxiliary a, EnergyBalanceExogenous ex)
    {
        //- Name: CanopyTemperature -Version: 1.0, -Time step: 1
        //- Description:
    //            * Title: CanopyTemperature Model
    //            * Author: Peter D. Jamieson, Glen S. Francis, Derick R. Wilson, Robert J. Martin
    //            * Reference: https://doi.org/10.1016/0168-1923(94)02214-5
    //            * Institution: New Zealand Institute for Crop and Food Research Ltd.,
    //            New Zealand Institute for Crop and Food Research Ltd.,
    //            New Zealand Institute for Crop and Food Research Ltd.,
    //            New Zealand Institute for Crop and Food Research Ltd.
    //        
    //            * ExtendedDescription: It is calculated from the crop heat flux and the boundary layer conductance 
    //            * ShortDescription: It is calculated from the crop heat flux and the boundary layer conductance 
        //- inputs:
    //            * name: minTair
    //                          ** description : minimum air temperature
    //                          ** datatype : DOUBLE
    //                          ** variablecategory : auxiliary
    //                          ** min : -30
    //                          ** max : 45
    //                          ** default : 0.7
    //                          ** unit : degC
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //                          ** inputtype : variable
    //            * name: maxTair
    //                          ** description : maximum air Temperature
    //                          ** datatype : DOUBLE
    //                          ** variablecategory : auxiliary
    //                          ** min : -30
    //                          ** max : 45
    //                          ** default : 7.2
    //                          ** unit : degC
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //                          ** inputtype : variable
    //            * name: cropHeatFlux
    //                          ** description : Crop heat flux
    //                          ** variablecategory : rate
    //                          ** inputtype : variable
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 10000
    //                          ** default : 447.912
    //                          ** unit : g/m**2/d
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //            * name: conductance
    //                          ** description : the boundary layer conductance
    //                          ** variablecategory : state
    //                          ** inputtype : variable
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 10000
    //                          ** default : 598.685
    //                          ** unit : m/d
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //            * name: lambdaV
    //                          ** description : latent heat of vaporization of water
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** default : 2.454
    //                          ** min : 0
    //                          ** max : 10
    //                          ** unit : MJ/kg
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //                          ** inputtype : parameter
    //            * name: rhoDensityAir
    //                          ** description : Density of air
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** default : 1.225
    //                          ** unit : kg/m**3
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //                          ** inputtype : parameter
    //            * name: specificHeatCapacityAir
    //                          ** description : Specific heat capacity of dry air
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** default : 0.00101
    //                          ** unit : MJ/kg/degC
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //                          ** inputtype : parameter
        //- outputs:
    //            * name: minCanopyTemperature
    //                          ** description : minimal Canopy Temperature  
    //                          ** datatype : DOUBLE
    //                          ** variablecategory : state
    //                          ** min : -30
    //                          ** max : 45
    //                          ** unit : degC
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //            * name: maxCanopyTemperature
    //                          ** description : maximal Canopy Temperature 
    //                          ** datatype : DOUBLE
    //                          ** variablecategory : state
    //                          ** min : -30
    //                          ** max : 45
    //                          ** unit : degC
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
        double minTair = a.minTair;
        double maxTair = a.maxTair;
        double cropHeatFlux = r.cropHeatFlux;
        double conductance = s.conductance;
        double minCanopyTemperature;
        double maxCanopyTemperature;
        minCanopyTemperature = minTair + (cropHeatFlux / (rhoDensityAir * specificHeatCapacityAir * conductance / lambdaV * 1000.0d));
        maxCanopyTemperature = maxTair + (cropHeatFlux / (rhoDensityAir * specificHeatCapacityAir * conductance / lambdaV * 1000.0d));
        s.minCanopyTemperature= minCanopyTemperature;
        s.maxCanopyTemperature= maxCanopyTemperature;
    }
}